File: src/test/java/cube8540/oauth/authentication/users/infra/UserExceptionTranslatorTest.java
package cube8540.oauth.authentication.users.infra;

import cube8540.oauth.authentication.error.message.ErrorMessage;
import cube8540.oauth.authentication.users.domain.UserAuthorizationException;
import cube8540.oauth.authentication.users.domain.UserInvalidException;
import cube8540.oauth.authentication.users.domain.UserNotFoundException;
import cube8540.oauth.authentication.users.domain.UserRegisterException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;

@DisplayName("유저 에러 변환기 테스트")
class UserExceptionTranslatorTest {

    private UserExceptionTranslator translator;

    @BeforeEach
    void setup() {
        this.translator = new UserExceptionTranslator();
    }

    @Test
    @DisplayName("UserNotFoundException 변환")
    void translateUserNotFoundException() {
        UserNotFoundException e = mock(UserNotFoundException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
    }

    @Test
    @DisplayName("UserRegisterException 변환")
    void translateUserRegisterException() {
        UserRegisterException e = mock(UserRegisterException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());
    }

    @Test
    @DisplayName("UserInvalidException")
    void translateUserInvalidException() {
        UserInvalidException e = mock(UserInvalidException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());
    }

    @Test
    @DisplayName("UserAuthorizationException")
    void translateUserAuthorizationException() {
        UserAuthorizationException e = mock(UserAuthorizationException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(e);
        assertEquals(HttpStatus.FORBIDDEN, response.getStatusCode());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/user/infra/UserExceptionTranslatorTest.kt
package cube8540.oauth.authentication.user.infra

import cube8540.oauth.authentication.users.domain.UserAuthorizationException
import cube8540.oauth.authentication.users.domain.UserInvalidException
import cube8540.oauth.authentication.users.domain.UserNotFoundException
import cube8540.oauth.authentication.users.domain.UserRegisterException
import cube8540.oauth.authentication.users.infra.UserExceptionTranslator
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpStatus
import java.util.*
import java.util.stream.Stream

class UserExceptionTranslatorTest {

    private val translator = UserExceptionTranslator()

    @ParameterizedTest
    @MethodSource(value = ["userNotFoundExceptions"])
    fun `translate user not found exception`(exception: UserNotFoundException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }

    @ParameterizedTest
    @MethodSource(value = ["userInvalidExceptions"])
    fun `translate user invalid exception`(exception: UserInvalidException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @ParameterizedTest
    @MethodSource(value = ["userRegisterExceptions"])
    fun `translate user register exception`(exception: UserRegisterException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @ParameterizedTest
    @MethodSource(value = ["userAuthorizationExceptions"])
    fun `translate user authorization exception`(exception: UserAuthorizationException) {
        val result = translator.translate(exception)
        assertThat(result.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
    }

    fun userNotFoundExceptions() = Stream.of(
        Arguments.of(UserNotFoundException.instance("user not found"))
    )

    fun userInvalidExceptions() = Stream.of(
        Arguments.of(UserInvalidException.instance(Collections.emptyList()))
    )

    fun userRegisterExceptions() = Stream.of(
        Arguments.of(UserRegisterException.existsIdentifier("exists identifier"))
    )

    fun userAuthorizationExceptions() = Stream.of(
        Arguments.of(UserAuthorizationException.alreadyCredentials("already credentials")),
        Arguments.of(UserAuthorizationException.invalidKey("invalid key")),
        Arguments.of(UserAuthorizationException.invalidPassword("invalid password")),
        Arguments.of(UserAuthorizationException.keyExpired("key expired"))
    )
}